### PR TITLE
Prefix dropdown in checkout doesn't select empty value by default if required

### LINKED ISSUE #18823: Prefix dropdown in checkout doesn't select empty value by default if required
### Preconditions (*)

1. Magento 2.2.6
2. Luma Theme

### Steps to reproduce (*)

1. In Admin Backend go to Stores > Configuration > Customers > Name and Address Options
2. Set "Show Prefix" to "Required"
3. Set "Prefix Dropdown Options" to ";Mr.;Mrs."
4. Save Config
5. Go to your store
6. Add a item to your cart
7. Go to checkout

### Expected result (*)

I expect the empty value in the prefix dropdown to be selected by default.
prefix_expected

### Actual result (*)

Instead of having the empty value selected by default it will be always the first value selected.
prefix_actual

### Additional information
issue_18823

### PR DESCRIPTION
**Pull request by engcom-Charlie**
### Related Pull Requests
Fixes #18823

### Fixed Issues (if relevant)

1. magento/magento2#18823:Prefix dropdown in checkout doesn't select empty value by default if required

### Manual testing scenarios (*)

Please see #18823

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: dev/tests/integration/testsuite/Magento/Customer/Model/OptionsTest.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    ): void {
+        $this->setConfig($showOptionConfig);
+        $this->setConfig($optionValuesConfig);
+
+        /** @var array $options */
+        $options = $optionType === self::STUB_OPTION_PREFIX_NAME
+            ? $this->model->getNamePrefixOptions()
+            : $this->model->getNameSuffixOptions();
+
+        $this->assertCount($expectedCount, $options);
```
Comment (orlangur) [inline dev/tests/integration/testsuite/Magento/Customer/Model/OptionsTest.php]: Excessive assert, please remove.

File: dev/tests/integration/testsuite/Magento/Customer/Model/OptionsTest.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+use Magento\Customer\Helper\Address;
+use Magento\Framework\App\Config\MutableScopeConfigInterface;
+use Magento\Framework\ObjectManagerInterface;
+use Magento\Store\Model\ScopeInterface;
+use Magento\TestFramework\Helper\Bootstrap;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Test for \Magento\Customer\Model\Options.
+ * @magentoDbIsolation enabled
```
Comment (orlangur) [inline dev/tests/integration/testsuite/Magento/Customer/Model/OptionsTest.php:21]: We don't need interaction with a real DB to check this logic, please convert to unit test.

File: dev/tests/integration/testsuite/Magento/Customer/Model/OptionsTest.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    private const XML_PATH_SUFFIX_OPTIONS = 'customer/address/suffix_options';
+    private const XML_PATH_PREFIX_SHOW = 'customer/address/prefix_show';
+    private const XML_PATH_PREFIX_OPTIONS = 'customer/address/prefix_options';
+
+    private const STUB_OPTION_PREFIX_NAME = 'prefix';
+    private const STUB_OPTION_SUFFIX_NAME = 'suffix';
+    private const STUB_CONFIG_VALUES = 'v1;v2';
+    private const STUB_CONFIG_VALUES_WITH_BLANK_OPTION = ';v1;v2';
+    private const STUB_EXPECTED_VALUES_WITH_BLANK_OPTION = [' ' => ' ', 'v1' => 'v1', 'v2' => 'v2'];
+    private const STUB_EXPECTED_VALUES = ['v1' => 'v1', 'v2' => 'v2'];
```
Comment (orlangur) [inline dev/tests/integration/testsuite/Magento/Customer/Model/OptionsTest.php]: Such test data is totally unreadable, please inline all constants and if you need some value in two places just create a variable in-place.

File: app/code/Magento/Customer/Model/Options.php [hunk 4]
```diff
@@ -97,12 +97,14 @@ private function prepareNamePrefixSuffixOptions($options, $isOptional = false)
         if (empty($options)) {
             return false;
         }
+
         $result = [];
-        $options = array_filter(explode(';', $options));
+        $options = explode(';', $options);
         foreach ($options as $value) {
-            $value = $this->escaper->escapeHtml(trim($value));
+            $value = $this->escaper->escapeHtml(trim($value)) ?: ' ';
```
Comment (orlangur) [inline app/code/Magento/Customer/Model/Options.php]: Why do we need a space value?

File: app/code/Magento/Customer/Model/Options.php [hunk 5]
```diff
@@ -97,12 +97,14 @@ private function prepareNamePrefixSuffixOptions($options, $isOptional = false)
...
             $result[$value] = $value;
         }
+
         if ($isOptional && trim(current($options))) {
             $result = array_merge([' ' => ' '], $result);
```
Comment (orlangur) [inline app/code/Magento/Customer/Model/Options.php]: Well, I don't understand the logic after such change. In optional field we add empty value, but in required field we allow empty value. How empty value can be valid for a required field?

Does it really works this way in registration?

https://github.com/magento/magento2/issues/18823#issuecomment-433606725

**Review by orlangur — CHANGES_REQUESTED** [review]
Let's first try to understand whether logic change is valid and only then apply necessary changes.

**Comment by engcom-Charlie** [comment]
Hi @orlangur, thank you for review, in the required field we add empty value only for it will be shown by default but we can't place an order with this on checkout or create a customer with blank value.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by orlangur** [comment]
@engcom-Charlie thanks, need to check this behavior in action.

**Comment by engcom-Charlie** [comment]
Hi @orlangur, do You've any updates?

**Comment by engcom-Charlie** [comment]
Hi @ihor-sviziev,
can You please review this PR?

**Review by VladimirZaets — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-7825](https://jira.corp.magento.com/browse/ENGCOM-7825) has been created to process this Pull Request
:eight_spoked_asterisk: @VladimirZaets, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @engcom-Charlie .

During testing, we faced the issue.

**Manual testing scenario:**

1. In Admin Backend go to Stores > Configuration > Customers > Name and Address Options;
2. Set "Show Prefix" to "Required";
3. Set "Prefix Dropdown Options" to ";Mr;Mrs;"
4. Save Config;
5. Go to your store;
6. Add an item to your cart;
7. Go to checkout;

As per [comment](https://github.com/magento/magento2/pull/26640#issuecomment-584193529), the expected behavior is:

`;Mr;Ms;`- select with `""`, `Mr`, `Ms` and `""`

but, **Actual Result:** :heavy_multiplication_x: `;Mr;Ms;`- select with `""`, `Mr`, `Ms`



@engcom-Charlie Could you take a look?

Thanks!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
Hi @ihor-sviziev, can You please review this PR?

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7825](https://jira.corp.magento.com/browse/ENGCOM-7825) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. In Admin Backend go to Stores > Configuration > Customers > Name and Address Options;
2. Set "Show Prefix" to "Required";
3. Set "Prefix Dropdown Options" to "`;Mr;Mrs`"
4. Save Config;
5. Go to your store;
6. Add an item to your cart;
7. Go to checkout;

**Before:** :heavy_multiplication_x: Instead of having the empty value selected by default it will be always the first value selected.



**After:** :heavy_check_mark: The empty value in the drop-down prefix was selected by default.



**Case 2. Set Prefix  to** **"**`Mr;Mrs;`**"**

**After:** :heavy_check_mark: Empty value at the end of the list



**Case 3. Set Prefix  to** **"**`;Mr;Ms;`**"**

**After:** :heavy_check_mark: Empty value at the beginning and end of the list



_Note:_ As per [comment](https://github.com/magento/magento2/pull/26640#issuecomment-584193529), the expected behavior is:

`Mr;Ms;`  - select with `Mr` , `Ms` and `""`
`;Mr;Ms` - select with `""`, `Mr` and `Ms`
`;Mr;Ms;`- select with `""`, `Mr`, `Ms` and `""`

**Pull request merged**